src/viztracer/modules/snaptrace.c

```diff
@@ -331,6 +334,17 @@ snaptrace_tracefunc(PyObject* obj, PyFrameObject* frame, int what, PyObject* arg
                 if (is_python && is_call && (frame->f_code->co_flags & CO_COROUTINE) != 0) {
                     info->paused = 1;
                     PyObject* curr_task = PyObject_CallObject(asyncio_tasks_current_task, NULL);
+                    if (!curr_task) {  // RuntimeError: no running event loop
```

Comment:
I think this piece should be reorganized. We could have future supports for other async libraries besides trio and the code will be messy.

I suggest to do a table based check for ``curr_task``:
```
PyObject* get_curr_task_list[] = {asyncio_tasks_current_task, trio_lowlevel_current_task};
PyObject* curr_task = NULL;
for (int i = 0; i < sizeof()/sizeof(); i++) {
    PyObject* get_curr_task = .. ;
    if (get_curr_task) {
    }
    if (curr_task) {
        break;
    } else {
    }
}
```
This would generate less duplicated code and it would be easier to add extra supports in the future.

You can assign ``curr_task`` to ``Py_None`` if you can't find any. That will stop the trace_func from searching for it in the sub-calls.

Reply:
Even though I implemented this I want to play devil's advocate that it might be premature generalization. For instance the obvious next candidate for implementing would be curio, but to get the current task from that backend you have to call an [async flavored function](https://github.com/dabeaz/curio/blob/master/curio/traps.py#L110-L114) so it won't work in this scheme. Also this array doesn't capture whether the task's name is in `task.name` `task.get_name()` or `task.id` or whatever. 🤷 Just thinking out loud.

Reply:
I'm not an expert in Python async, but I don't think it's possible to support an async flavored function in tracefunc. It has to be a fast sync function.

I don't think you need to generalize them to be a specific function. As you can tell, the two functions are almost identical. You can put the code in the for loop, which is short enough that won't create too much confusion, and could save a function call. You can check if ``asyncio_tasks_current_task`` is ``NULL`` even though it's never true. Just to make the code cleaner.

Putting the array as global variable is fine, and I think it should be ``static const``.

Reply:
see 1232a19, does it look better?